willow-react-demo/src/stores/ChatGptStore/index.tsx

```diff
@@ -0,0 +1,130 @@
+import { makeObservable, action, observable, computed } from "mobx";
+import OpenAI from 'openai';
+import { ttsPlayer } from '../../widgets/tts';
+import { isPhraseComplete } from "./helpers";
+
+export enum AUTHOR {
+  USER = 'user',
+  GPT = 'gpt',
+}
+
+export interface IMessage {
+  author: AUTHOR,
+  text: string;
+}
+
+export class ChatGptStore {
+  openai: OpenAI;
+
+  messages: IMessage[] = [];
+
+  activeGptOutput: string = '';
+  activeUserInput: string = '';
+
+  isLoading: boolean = false;
+
+  constructor() {
+    makeObservable(this, {
+      isLoading: observable,
+      messages: observable,
+      activeGptOutput: observable,
+      activeUserInput: observable,
+      setUserInput: action,
+      setGptOutput: action,
+      loadGptAnswer: action,
+      clearMessages: action,
+      loadMessages: action
+    })
+
+    this.openai = new OpenAI({
+      apiKey: String(process.env.REACT_APP_SECRET_OPEN_AI),
+      dangerouslyAllowBrowser: true
+    });
+
+    this.loadMessages();
+  }
+
+  setUserInput = (text: string) => {
+    this.activeUserInput = text;
+
+    this.messages.push({
+      author: AUTHOR.USER,
+      text
+    })
+  }
+
+  setGptOutput = (text: string) => {
+    this.activeGptOutput = text;
+  }
+
+  loadGptAnswer = async () => {
+    this.isLoading = true;
+
+    try {
+      const content = this.activeUserInput;
+      let resMessage = ''
+
+      ttsPlayer.clear();
+      ttsPlayer.play();
+
+      const stream = await this.openai.chat.completions.create({
+        model: 'gpt-4',
+        messages: [{ role: 'user', content }],
+        stream: true,
+        max_tokens: 100,
+        temperature: 0.8
+      });
+
+      const lines: string[] = [];
+      let currentLineIdx = 0;
+      let text = '';
+
+      for await (const part of stream) {
+        text = (part.choices[0]?.delta?.content || '');
+
+        resMessage += text;
+        this.setGptOutput(resMessage);
+        lines[currentLineIdx] = (lines[currentLineIdx] ?? '') + text;
+
+        if (isPhraseComplete(lines[currentLineIdx], !currentLineIdx)) {
+          ttsPlayer.setText(lines, false);
```

Comment:
seems to be constantly overwriting the tts player's text buffer - is it the right streaming approach? Does ttsPlayer wait until complete flag is set to true? https://github.com/harmony-one/x/blob/9ef2ff04e15973aca95ef71777d4f6181a06d5c1/willow-react-demo/src/widgets/tts/types.ts#L28